man/dioddate.8.in
.TH dioddate 8 "@PACKAGE_VERSION@" "@PACKAGE_NAME@" "@PACKAGE_NAME@"
.SH NAME
dioddate \- get server system time
.SH SYNOPSIS
\fBdioddate\fR \fI[OPTIONS] [-s NAME] [file [file...]]\fR
.SH DESCRIPTION
.B dioddate
obtains the current system time and zone information from a \fBdiod\fR
server by reading \fIctl:date\fR.  If the \fI--set-time\fR option
is specified, sets the local system time on the client; otherwise
prints the server time on stdout.
.LP
This is a simple way to obtain rough time synchronization before
issuing a MUNGE-authenticated attach request, for example in a bootstrap
environment where the "ctl" synthetic file system is exported with
the "noauth" option.
.SH OPTIONS
.TP
.I "-s, --server NAME"
The server in IP[:PORT], HOST[:PORT], or /path/to/socket form
(default localhost:564).
.TP
.I "-m, --msize SIZE"
The maximum request size including 9P headers (default 65536).
.TP
.I "-u, --uid UID"
Try to attach to the server as the specified user (default your effective uid).
.TP
.I "-t, --timeout SECS"
Force timeout after specified number of seconds (default no timeout).
.TP
.I "-S, --set-time"
Set the local system time to match the server's time.
.SH "SEE ALSO"
diod (8)


NEWS
=========================================================================
Release Notes for DIOD version 1.0.24                         31 Mar 2015
=========================================================================

* Replace sysv-init script with systemd unit file

* Move fully to github (including docs)

* Misc. minor updates, mostly tests and build environment, for rhel7

=========================================================================
Release Notes for DIOD version 1.0.23                         26 Aug 2014
=========================================================================

* Add dist tag to release (TOSS-2662)

* Minor autotools forward-portability (issue 124)

* Doc: AUTHORS, README, license headers updated

* Stop tracking additional autotools build products.

=========================================================================
Release Notes for DIOD version 1.0.22                         19 Aug 2014
=========================================================================

* Stop tracking autotools build products in git.

* Fix assertion failure at srv.c:754 (issue 122)

* Portability fixes for musl libc (Felix Janda)

* Change default rdma port to 5640 to match kernel

* rdmatrans: only register the used buffer to the network
  adapter (Dominique Martinet)

* fix configure --with-tcmalloc logic

* libnpfs: fids cloned in xattrwalk should inherit flags

* rdmatrans: compilation error

=========================================================================
Release Notes for DIOD version 1.0.21                         31 May 2013
=========================================================================

* Require that "ctl" be explicitly exported.  This allows export options
  such as "ro", "noauth", and "privport" to apply to the ctl synthetic
  file system.

* The "auth_required_ctl" config option introduced in 1.0.15 is now
  deprecated.

* Add test coverage for trusted.* and security.* xattr namespaces.

* Add more mmap test coverage.

=========================================================================
Release Notes for DIOD version 1.0.20                         17 May 2013
=========================================================================

* Fix diodshowmount segfault on exit with 'noauth'.

* Add diodls --privport option.

* Additional test coverage (atomic create test, pathwalk test)

=========================================================================
Release Notes for DIOD version 1.0.19                         07 May 2013
=========================================================================

* Additional test coverage for xattrs:
  - Verify that setxattr flags XATTR_CREATE, XATTR_REPLACE work.
  - Verify that that removal of a nonexistent user xattr returns ENODATA

* Fix failure of tests/user/t17 when SELinux is enabled (bug in the test)

=========================================================================
Release Notes for DIOD version 1.0.18                         06 May 2013
=========================================================================

* Man page update.

=========================================================================
Release Notes for DIOD version 1.0.17                         06 May 2013
=========================================================================

* RPM: add missing BuildRequires for 'libattr-devel'.

=========================================================================
Release Notes for DIOD version 1.0.15                         03 May 2013
=========================================================================

* Add xattr support (issue 95)

* Support for dracut bootstrap of 9nbd network block device:
  - New 'dioddate' utility to synchronize system time with server's,
    and 'auth_required_ctl' config option which allows MUNGE
    authentication to be disabled for the ctl synthetic file system.
  - Add 'noauth' and 'privport' export options which disable MUNGE auth,
    and require attaches to originate from a privileged port (512-1024)
    for a specific export.    

=========================================================================
Release Notes for DIOD version 1.0.14                         12 Jul 2012
=========================================================================

* Add mount.diod --attach and --detach options to support 9nbd
  (9P network block device).

* Allow non-directories to be exported (issue 113)

* Return Rlerror(ENOENT) in response to Tauth if auth not required
  instead of Rlerror(0), which confused kernel 9P code.
    
=========================================================================
Release Notes for DIOD version 1.0.13                         22 May 2012
=========================================================================

* Allow DIOD_MOUNTOPTS in /etc/sysconfig/auto.diod

  Normalize -s,--server NAME option in diodcat, diodload, diodls,
  and diodshowmount utilities, where NAME can be IP[:PORT], HOST[:PORT],
  or /path/to/socket.

* Add diodshowmount utility (issue 111)

* Add diodls utility

* Debugging/cleanup for issue 108
  - refactor and clarify error paths in ioctx open/close functions
  - normalize error messages when ioctx state is not as expected

=========================================================================
Release Notes for DIOD version 1.0.12                         15 May 2012
=========================================================================

* Rework fdtrans somewhat for clarity (issue 109)

* Test improvements:
  - misc/t15: don't use munge auth or usleep (caused hang on build farm)
  - misc/t15: update valgrind suppressions for 32b Ubuntu tests

=========================================================================
Release Notes for DIOD version 1.0.11                         14 May 2012
=========================================================================

* Return EINVAL when a file is opened with O_DIRECT flag (issue 110).

* Drop unused maxmmap feature.


src/cmd/dioddate.c
/************************************************************\
 * Copyright 2010 Lawrence Livermore National Security, LLC
 * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
 *
 * This file is part of the diod 9P server project.
 * For details, see https://github.com/chaos/diod.
 *
 * SPDX-License-Identifier: GPL-2.0-or-later
\************************************************************/

/* dioddate.c - sync time with server */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <time.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <stdarg.h>
#include <stdio.h>
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <string.h>
#include <errno.h>
#include <ctype.h>
#include <libgen.h>
#include <signal.h>

#include "src/libnpfs/9p.h"
#include "src/libnpfs/npfs.h"
#include "src/libnpclient/npclient.h"

#include "src/liblsd/list.h"
#include "src/libdiod/diod_log.h"
#include "src/libdiod/diod_sock.h"
#include "src/libdiod/diod_auth.h"

#define OPTIONS "s:m:u:t:S"
#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long (ac,av,opt,lopt,NULL)
static const struct option longopts[] = {
    {"server",  required_argument,      0, 's'},
    {"msize",   required_argument,      0, 'm'},
    {"uid",     required_argument,      0, 'u'},
    {"timeout", required_argument,      0, 't'},
    {"set-time",no_argument,            0, 'S'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt (ac,av,opt)
#endif

static void sigalarm (int arg);

static void
usage (void)
{
    fprintf (stderr,
"Usage: dioddate [OPTIONS] [-s HOST[:PORT]]\n"
"   -S,--set-time         set local time to match server's\n"
"   -s,--server HOST:PORT server (default localhost:564)\n"
"   -m,--msize            msize (default 65536)\n"
"   -u,--uid              authenticate as uid (default is your euid)\n"
"   -t,--timeout SECS     give up after specified seconds\n"
);
    exit (1);
}

int
main (int argc, char *argv[])
{
    char *aname = "ctl";
    char *server = NULL;
    int msize = 65536;
    uid_t uid = geteuid ();
    int topt = 0;
    int Sopt = 0;
    int flags = 0;
    int fd, c;
    Npcfsys *fs = NULL;
    Npcfid *afid = NULL, *root = NULL;
    char *buf = NULL;
    struct timeval tv;
    struct timezone tz;

    diod_log_init (argv[0]);

    opterr = 0;
    while ((c = GETOPT (argc, argv, OPTIONS, longopts)) != -1) {
        switch (c) {
            case 's':   /* --server HOST[:PORT] or /path/to/socket */
                server = optarg;
                break;
            case 'm':   /* --msize SIZE */
                msize = strtoul (optarg, NULL, 10);
                break;
            case 'u':   /* --uid UID */
                uid = strtoul (optarg, NULL, 10);
                break;
            case 't':   /* --timeout SECS */
                topt = strtoul (optarg, NULL, 10);
                break;
            case 'S':   /* --set-time */
                Sopt = 1;
                break;
            default:
                usage ();
        }
    }

    if (signal (SIGPIPE, SIG_IGN) == SIG_ERR)
        err_exit ("signal");
    if (signal (SIGALRM, sigalarm) == SIG_ERR)
        err_exit ("signal");

    if (topt > 0)
        alarm (topt);

    if ((fd = diod_sock_connect (server, flags)) < 0)
        exit (1);

    if (!(fs = npc_start (fd, fd, msize, 0))) {
        errn (np_rerror (), "error negotiating protocol with server");
        goto done;
    }
    if (!(afid = npc_auth (fs, aname, uid, diod_auth)) && np_rerror () != 0) {
        errn (np_rerror (), "error authenticating to server");
        goto done;
    }
    if (!(root = npc_attach (fs, afid, aname, uid))) {
        errn (np_rerror (), "error attaching to aname='%s'", aname);
        goto done;
    }
    buf = npc_aget (root, "date");
    if (!buf) {
        errn (np_rerror (), "error reading date");
        goto done;
    }

    int64_t sec = 0, usec = 0;
    if (sscanf (buf, "%"SCNd64".%"SCNd64" %d.%d", &sec, &usec,
                                    &tz.tz_minuteswest, &tz.tz_dsttime) != 4) {
        msg ("error scanning returned date: %s", buf);
        goto done;
    }
    tv.tv_sec = sec;
    tv.tv_usec = usec;

    if (Sopt) {
        if (settimeofday (&tv, &tz) < 0)
            err_exit ("settimeofday");
    } else {
        time_t t = tv.tv_sec;
        printf ("%s", ctime (&t));
    }
done:
    if (buf)
        free (buf);
    if (root && npc_clunk (root) < 0)
        errn_exit (np_rerror (), "error clunking %s", aname);
    if (afid && npc_clunk (afid) < 0)
        errn_exit (np_rerror (), "error clunking afid");
    if (fs)
        npc_finish (fs);
    close (fd);

    diod_log_fini ();

    exit (0);
}

static void sigalarm (int arg)
{
    msg_exit ("timed out");
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
